eclipse-theia/theia
# Advance Debugging

#### What it does

- [x] https://github.com/theia-ide/theia/issues/5769 - watch expressions
  - [x] a CQ for add icon: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20445 (resolved)
- [x] https://github.com/theia-ide/theia/issues/5768 - exception breakpoints
- [x] https://github.com/theia-ide/theia/issues/5767 - function breakpoints
  - [x] a CQ for function breakpoint icon: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20470
- [x] https://github.com/theia-ide/theia/issues/3807 - inlined breakpoints
- [x] https://github.com/theia-ide/theia/issues/5788 - fix identify of breakpoints in the plugin process
- [x] add copyrights to all copied icons

Pause on an exception with watch expressions:
<img width="1792" alt="Screen Shot 2019-07-24 at 09 45 04" src="https://user-images.githubusercontent.com/3082655/61774747-d430b500-adf7-11e9-90f8-7a174ad6baaf.png">

#### How to test

- watch expressions: add watch expressions and start debugging, they should be updated as you go
- exception breakpoints:
  - they are provided by a debug adapter on the first run, so start debugging something
  - after that you can enable them
  - after the debug session is finished, such breakpoints should be preserved anyway and reevaluated on next run
  - when exception breakpoint hits you should be able to see the reason in an editor
- function breakpoints:
  - install nodejs vscode extensions, https://marketplace.visualstudio.com/items?itemName=ms-vscode.node-debug2 and https://marketplace.visualstudio.com/items?itemName=ms-vscode.node-debug
  - uninstall native Theia nodejs extension from example app and rebuild it
  - follow https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_function-breakpoints
    - it works only with legacy protocol for Node.js < 8 and only for global functions which were already defined at runtime
![function_breakpoint](https://user-images.githubusercontent.com/3082655/61804342-0e1fac80-ae34-11e9-801a-e02bbd5786fa.gif)

- inlined breakpoints
  - look out for regressions with usual breakpoints, one should be able to mix them, disable, enable, edit on the same line via the context menu in the ruler and the breakpoints debug view

- breakpoints identity issues in the plugin host process: TBD
